// @clr/angular/esm5/accordion/models/accordion.model.js
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { AccordionStrategy } from '../enums/accordion-strategy.enum';
import { AccordionStatus } from '../enums/accordion-status.enum';
var accordionCount = 0;
var AccordionPanelModel = /** @class */ (function () {
    function AccordionPanelModel(id, accordionId) {
        this.id = id;
        this.accordionId = accordionId;
        this.status = AccordionStatus.Inactive;
        this.index = null;
        this.disabled = false;
        this.open = false;
        this.templateId = this.id + "-" + this.accordionId;
    }
    return AccordionPanelModel;
}());
export { AccordionPanelModel };
var AccordionModel = /** @class */ (function () {
    function AccordionModel() {
        this.strategy = AccordionStrategy.Default;
        this.accordionCount = accordionCount++;
        this._panels = {};
    }
    Object.defineProperty(AccordionModel.prototype, "panels", {
        get: function () {
            var _this = this;
            return Object.keys(this._panels).map(function (id) { return _this._panels[id]; });
        },
        enumerable: true,
        configurable: true
    });
    AccordionModel.prototype.setStrategy = function (strategy) {
        this.strategy = strategy;
    };
    AccordionModel.prototype.updatePanelOrder = function (ids) {
        var _this = this;
        ids.forEach(function (id, index) { return (_this._panels[id].index = index); });
        this.removeOldPanels(ids);
    };
    AccordionModel.prototype.addPanel = function (id, open) {
        if (open === void 0) { open = false; }
        this._panels[id] = new AccordionPanelModel(id, this.accordionCount);
        this._panels[id].open = open;
    };
    AccordionModel.prototype.togglePanel = function (panelId, open) {
        var panelIsOpen = this._panels[panelId].open;
        if (this.strategy === AccordionStrategy.Default) {
            this.closeAllPanels();
        }
        this._panels[panelId].open = open !== undefined ? open : !panelIsOpen;
    };
    AccordionModel.prototype.disablePanel = function (panelId, disabled) {
        this._panels[panelId].disabled = disabled;
    };
    AccordionModel.prototype.closeAllPanels = function () {
        var _this = this;
        this.panels.forEach(function (panel) { return (_this._panels[panel.id].open = false); });
    };
    AccordionModel.prototype.removeOldPanels = function (ids) {
        var _this = this;
        this.panels
            .filter(function (panel) { return ids.find(function (id) { return id === panel.id; }) === undefined; })
            .forEach(function (panel) { return delete _this._panels[panel.id]; });
    };
    return AccordionModel;
}());
export { AccordionModel };
