// @clr/core/common/mixins/css-helpers.js
/*
 * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
export class CssHelpers {
    hasClassname(text) {
        return this.classList.contains(text);
    }
    addClassname(text) {
        this.classList.add(text);
    }
    removeClassname(text) {
        this.classList.remove(text);
    }
    removeClassnamesUnless(classnamesToRemove, classnamesToKeep) {
        classnamesToRemove.forEach(removeMe => {
            if (classnamesToKeep.indexOf(removeMe) < 0) {
                this.classList.remove(removeMe);
            }
        });
    }
    removeClassnames(classnamesToRemove) {
        this.removeClassnamesUnless(classnamesToRemove, []);
    }
    updateEquilateralStyles(size) {
        this.style.width = size || '';
        this.style.height = size || '';
    }
    removeEquilateralStyles() {
        this.updateEquilateralStyles();
    }
    addEquilateralStyles(size) {
        this.updateEquilateralStyles(size);
    }
}
