// @clr/angular/esm5/utils/focus/focus-on-view-init/focus-on-view-init.module.js
import { __decorate } from "tslib";
/*
 * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ClrFocusOnViewInit } from './focus-on-view-init';
import { FOCUS_ON_VIEW_INIT_PROVIDER } from './focus-on-view-init.provider';
export var FOCUS_ON_VIEW_INIT_DIRECTIVES = [ClrFocusOnViewInit];
var ClrFocusOnViewInitModule = /** @class */ (function () {
    function ClrFocusOnViewInitModule() {
    }
    ClrFocusOnViewInitModule = __decorate([
        NgModule({
            imports: [CommonModule],
            declarations: [FOCUS_ON_VIEW_INIT_DIRECTIVES],
            providers: [FOCUS_ON_VIEW_INIT_PROVIDER],
            exports: [FOCUS_ON_VIEW_INIT_DIRECTIVES],
        })
    ], ClrFocusOnViewInitModule);
    return ClrFocusOnViewInitModule;
}());
export { ClrFocusOnViewInitModule };
